Asynchronous NIF Rewrite

potatosalad opened the pull request:
Using [gburd/lmdb](https://github.com/gburd/lmdb) and [basho-labs/wterl](https://github.com/basho-labs/wterl) as a reference, I rewrote this NIF to use [async_nif.h](https://gist.github.com/gburd/4121795) so that it is _less_ blocking than the original synchronous version.

It should be a drop-in replacement to the 0.2.x version.  I added an additional function, `ekstat:close/1`, for forcing [kstat_close(3kstat)](https://www.illumos.org/man/3KSTAT/kstat_close) to be called.

All of the tests are passing for me on SmartOS joyent_20140221T042147Z with Erlang R16B02.

I'm hoping to do something similar to [project-fifo/erltrace](https://github.com/project-fifo/erltrace) when I have the time.

Hopefully you find this useful!

**Edit 2014/03/05:** I forgot to mention, the biggest change in this version in terms of backwards compatibility are the error messages.  If there's an OOM error, for example, instead of returning (for example) `{error, {kstat, "ks_nvpair_t malloc"}}` it will now return `{error, {enomem, "Not enough space"}}`.  The error checking is much more thorough now as well compared to the previous version.

potatosalad committed:
Fix a tiny issue when compiling ekstat.c with -Wmaybe-uninitialized.

potatosalad committed:
Version 0.2.2

potatosalad committed:
Experimenting with a asynchronous NIF version of ekstat.

potatosalad committed:
There's still a memory leak inside ekstat:update/1

potatosalad committed:
Finish up the asynchronous NIF version of ekstat.

potatosalad committed:
Version 0.3.0

Licenser commented:
wow I fell seriously bad about this, I missed this PR entirely :( I'll review it this week

potatosalad committed:
Merge remote-tracking branch 'Licenser/master' into develop

elvisci commented on src/ekstat.erl:
```diff
@@ -20,38 +26,76 @@
+read(Handle, Class) ->
+    ?ASYNC_NIF_CALL(fun read_nif/3, [Handle, Class]).
+
+-spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern()) -> [kstat_row()] | kstat_error().
```
Line 51 is too long: -spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern()) -> [kstat_row()] | kstat_error()..

elvisci commented on src/ekstat.erl:
```diff
@@ -20,38 +26,76 @@
+read(Handle, Class, Module, Instance) ->
+    ?ASYNC_NIF_CALL(fun read_nif/5, [Handle, Class, Module, Instance]).
+
+-spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern(), Name::kstat_pattern()) -> [kstat_row()] | kstat_error().
```
Line 59 is too long: -spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern(), Name::kstat_pattern()) -> [kstat_row()] | kstat_error()..

elvisci commented on src/ekstat.erl:
```diff
@@ -1,14 +1,20 @@
 -type kstat_pattern() :: '_' | integer() | string().
--type kstat_error() :: {error, {kstat, string()}}.
+-type kstat_error() :: {error, atom()} | {error, {atom(), string()}}.
 -type kstat_statistic() :: {Statistic::string(), Value::integer()} | {Statistic::string(), Value::string()}.
```
Line 18 is too long: -type kstat_statistic() :: {Statistic::string(), Value::integer()} | {Statistic::string(), Value::string()}..

elvisci commented on src/ekstat.erl:
```diff
@@ -20,38 +26,76 @@
+read(Handle, Class, Module, Instance, Name) ->
+    ?ASYNC_NIF_CALL(fun read_nif/6, [Handle, Class, Module, Instance, Name]).
+
+-spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern(), Name::kstat_pattern(), Statistic::kstat_pattern()) -> [kstat_row()] | kstat_error().
```
Line 63 is too long: -spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern(), Name::kstat_pattern(), Statistic::kstat_pattern()) -> [kstat_row()] | kstat_error()..

elvisci commented on src/ekstat.erl:
```diff
@@ -20,38 +26,76 @@
+read(Handle) ->
+    ?ASYNC_NIF_CALL(fun read_nif/2, [Handle]).
+
+-spec read(Handle::binary(), Class::kstat_pattern()) -> [kstat_row()] | kstat_error().
```
Line 47 is too long: -spec read(Handle::binary(), Class::kstat_pattern()) -> [kstat_row()] | kstat_error()..

elvisci commented on src/ekstat.erl:
```diff
@@ -1,14 +1,20 @@
--type kstat_error() :: {error, {kstat, string()}}.
+-type kstat_error() :: {error, atom()} | {error, {atom(), string()}}.
 -type kstat_statistic() :: {Statistic::string(), Value::integer()} | {Statistic::string(), Value::string()}.
 -type kstat_row() :: {Class::string(), Module::string(), Instance::integer(), Name::string(), [kstat_statistic()]}.
```
Line 19 is too long: -type kstat_row() :: {Class::string(), Module::string(), Instance::integer(), Name::string(), [kstat_statistic()]}..

elvisci commented on src/ekstat.erl:
```diff
@@ -20,38 +26,76 @@
+
+-spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern(), Name::kstat_pattern(), Statistic::kstat_pattern()) -> [kstat_row()] | kstat_error().
+read(Handle, Class, Module, Instance, Name, Statistic) ->
+    ?ASYNC_NIF_CALL(fun read_nif/7, [Handle, Class, Module, Instance, Name, Statistic]).
```
Line 65 is too long:     ?ASYNC_NIF_CALL(fun read_nif/7, [Handle, Class, Module, Instance, Name, Statistic])..

elvisci commented on src/ekstat.erl:
```diff
@@ -20,38 +26,76 @@
+read(Handle, Class, Module) ->
+    ?ASYNC_NIF_CALL(fun read_nif/4, [Handle, Class, Module]).
+
+-spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern()) -> [kstat_row()] | kstat_error().
```
Line 55 is too long: -spec read(Handle::binary(), Class::kstat_pattern(), Module::kstat_pattern(), Instance::kstat_pattern()) -> [kstat_row()] | kstat_error()..

elvisci commented on src/ekstat.erl:
```diff
@@ -1,14 +1,20 @@
 
 %% API
--export([open/0, update/1, clear/1, read/1, read/2, read/3, read/4, read/5, read/6]).
+-export([open/0, close/1, clear/1, update/1, read/1, read/2, read/3, read/4, read/5, read/6]).
```
Line 4 is too long: -export([open/0, close/1, clear/1, update/1, read/1, read/2, read/3, read/4, read/5, read/6])..

elvisci commented on src/ekstat.erl:
```diff
@@ -1,65 +1,162 @@
+        Name :: kstat_pattern(),
+        Statistic :: kstat_pattern().
+read(Handle, Class, Module, Instance, Name, Statistic) ->
+    ?ASYNC_NIF_CALL(fun read_nif/7, [Handle, Class, Module, Instance, Name, Statistic]).
```
Line 115 is too long:     ?ASYNC_NIF_CALL(fun read_nif/7, [Handle, Class, Module, Instance, Name, Statistic])..

potatosalad committed:
Fix formatting errors so elvis doesn't yell at us.

potatosalad commented:
Hey @Licenser, I'd forgotten about this, too.

I corrected the merge conflicts from master and fixed the formatting issues that elvis was mad about.

Everything still compiles and runs just fine on my local machine running `joyent_20160428T170316Z`.

I apologize for this pull request including a version bump, my git etiquette was much more crude 2 years ago ¯\\\_(ツ)\_/¯

potatosalad committed:
4740 kstat(1M) prints wrong snaptime

https://github.com/joyent/illumos-joyent/commit/6a19b8666104dbf912da5a4fb193c9c2b29e21bb

potatosalad commented:
I also incorporated the fix from the upstream kstat(1M) from a few months ago which was causing snaptime to be inaccurate: https://github.com/joyent/illumos-joyent/commit/6a19b8666104dbf912da5a4fb193c9c2b29e21bb

Licenser commented:
Awesome, this is a huuuge pice of work! Much appriciated :D

Licenser commented:
if you ever stumble by in berlin let me know I owe you a coffee :)
